draft-tus-resumable-uploads-protocol.md

```diff
@@ -188,29 +188,34 @@ If the end of the request body is not the end of the upload, the `Upload-Incompl
 
 The client MUST NOT perform multiple Upload Transfer Procedures ({{upload-transfer}}) for the same token in parallel to avoid race conditions and data loss or corruption. The server is RECOMMENDED to take measures to avoid parallel Upload Transfer Procedures: The server MAY terminate any ongoing Upload Transfer Procedure ({{upload-transfer}}) for the same token. Since the client is not allowed to perform multiple transfers in parallel, the server can assume that the previous attempt has already failed. Therefore, the server MAY abruptly terminate the previous HTTP connection or stream.
 
-If the offset in the `Upload-Offset` header field does not match the value 0, the offset provided by the immediate previous Offset Retrieving Procedure ({{offset-retrieving}}), or the end offset of the immediate previous incomplete transfer, the server MUST respond with `409 (Conflict)` status code.
+If the offset in the `Upload-Offset` header field does not match the value 0, the offset provided by the immediate previous Offset Retrieving Procedure ({{offset-retrieving}}), or the end offset of the immediate previous incomplete transfer, the server MUST respond with `409 (Conflict)` status code and the `Upload-Offset` header set to the expected upload resumption offset.
 
-If the request completes successfully and the entire upload is complete, the server MUST acknowledge it by responding with a successful status code between 200 and 299 (inclusive). Server is RECOMMENDED to use `201 (Created)` response if not otherwise specified. The response MUST NOT include the `Upload-Incomplete` header with the value of true.
+If the request completes successfully and the entire upload is complete, the server MUST acknowledge it by responding with a successful status code between 200 and 299 (inclusive) and the `Upload-Offset` header set to the full upload's length. The Server is RECOMMENDED to use `201 (Created)` response if not otherwise specified. The response MUST NOT include the `Upload-Incomplete` header with the value of true.
 
 If the request completes successfully but the entire upload is not yet complete indicated by the `Upload-Incomplete` header, the server MUST acknowledge it by responding with the `201 (Created)` status code, the `Upload-Incomplete` header set to true, and the `Upload-Offset` header set to the new upload resumption offset.
 
+The Client MAY use the `Upload-Offset` header in the response to verify that the Server received the expected number of bytes. If the response does not contain the `Upload-Offset` header, and the status code is not `404 (Not Found)`, the Client MUST perform the Offset Retrieving Procedure ({{offset-retrieving}}) to retrieve the expected offset before continuing the upload.
```

Comment:
> If the offset in the `Upload-Offset` header field does not match the value 0, the offset provided by the immediate previous Offset Retrieving Procedure ({{offset-retrieving}}), or the end offset of the immediate previous incomplete transfer, the server MUST respond with `409 (Conflict)` status code

I think this is slightly in conflict with the statement above. Client should know the offset already, so it should never rely on the `Upload-Offset` in Upload Transfer Procedure responses. The client MAY check if the `Upload-Offset` in the response matches expectations.

Reply:
I don't think I really follow, could you give more details on how it's in conflict? The added section describes what should happen if the `Upload-Offset` is not present _at all_ (e.g. a proxy stripped it) and was added to prevent the client from assuming that that a missing header is the same as zero. For anything but the 404 response, this means that the client must perform an offset retrieving procedure to get the correct offset. For a 404 there is obviously no need as the upload does not exist.

The client may check if the `Upload-Offset` matches what it expects and if not it may abort the upload or perform a new offset retrieving procedure to get the real offset and continue the upload from there. If the `Upload-Offset` does match then there is no issue. 

> or the end offset of the immediate previous incomplete transfer,

This should be updated to remove the word "incomplete" as this is no longer the case.

Reply:
> If the response does not contain the `Upload-Offset` header, and the status code is not `404 (Not Found)`, the Client MUST perform the Offset Retrieving Procedure ({{offset-retrieving}}) to retrieve the expected offset before continuing the upload.

This sentence implies that if the response contains `Upload-Offset`, client can use it to continue the upload. I think this is backwards. Client should always know the expected `Upload-Offset` here already, and there is something seriously broken if client doesn't. The sentence quoted in my previous comment does not allow the offset in this response to be used.